crates/host/src/wasmbus/mod.rs

```diff
@@ -1977,14 +1982,34 @@ impl Host {
 
         let annotations = annotations.map(|annotations| annotations.into_iter().collect());
         let mut providers = self.providers.write().await;
-        let hash_map::Entry::Occupied(mut entry) = providers.entry(provider_ref) else {
+        let hash_map::Entry::Occupied(mut entry) = providers.entry(provider_ref.clone()) else {
             return Ok(SUCCESS.into());
         };
         let provider = entry.get_mut();
         let instances = &mut provider.instances;
         if let hash_map::Entry::Occupied(entry) = instances.entry(link_name.clone()) {
             if entry.get().annotations == annotations {
                 let ProviderInstance { id, child, .. } = entry.remove();
+
+                // Send a request to the provider, requesting a graceful shutdown
+                if let Ok(payload) = serde_json::to_vec(&json!({ "host_id": host_id })) {
+                    if let Err(e) = self
+                        .nats
+                        .send_request(
+                            format!(
+                                "wasmbus.rpc.{}.{provider_ref}.{link_name}.shutdown",
+                                self.host_config.lattice_prefix
+                            ),
+                            async_nats::Request::new()
+                                .payload(payload.into())
+                                .timeout(self.host_config.provider_shutdown_delay),
+                        )
+                        .await
+                    {
+                        warn!(?e, "Provider didn't gracefully shut down in time, shutting down forcefully");
+                    }
+                }
+
                 child.abort();
```

Comment:
Should we wait for `provider_shutdown_delay` before sending this? Does an issue exist for that yet?

Reply:
We only have to wait for `provider_shutdown_delay` in terms of waiting for a _reply_ from the capability provider, we don't always have to wait for that time if it only takes, say, 100ms for the provider to cleanup. So having it as the request timeout feels correct, unless I'm thinking about it wrong 🤔

Reply:
Ah okay. I wasn't sure if it was possible for a provider to immediately ack-ed the shutdown message before actually cleaning up its resources/shutting down gracefully

Reply:
Hm, yeah I think it is possible for providers to do that but we wouldn't know if they did or not 🤔